### ISSUE TITLE
Query Batching support?

### ISSUE THREAD

**Issue by acronoah** [graphql-dotnet/server#122]
I turned on Apollo client batching this morning and was surprised to find that GraphQLHttpMiddleware doesn't seem to have any support for it.  So I added it.

Here's the [commit](https://github.com/graphql-dotnet/server/commit/90b637750a4326bf846c227a5ec82ba356ba5426).

I can create a pull request in a one-off branch if you'd like.

**Comment by joemcbride** [comment]
I'd certainly be interested in seeing that added.

**Comment by pekkah** [comment]
@acronoah 👍

**Comment by acronoah** [comment]
Work has been consuming the bulk of my waking hours as of late, but I will make an effort in the next week to get you a PR.

**Comment by KreativJos** [comment]
@acronoah What is the status of your PR? We implemented a batch-request too, but it would be nice if it was supported from the gql-dotnet framework itself.

**Comment by ZenSoftware** [comment]
@acronoah I have been working on trying to get Apollo Query Batching to work as well.   Though, I am not really sure what the standard GQL batched response should be.  See the following bare-bones Web API endpoint:

Any ideas would really help @joemcbride !

```C#
public async Task<HttpResponseMessage> PostAsync(GraphQLQuery[] queries)
{
	ExecutionResult[] results = await Task.WhenAll(
		queries.Select(query => _executer.ExecuteAsync(options =>
		{
			var variables = query.Variables as Newtonsoft.Json.Linq.JObject;

			options.Schema = (Schema)schema;
			options.Query = query.Query;
			options.OperationName = query.OperationName;
			options.Inputs = variables.ToInputs();
		}))
	);

	HttpStatusCode httpResult = results.Any(result => result.Errors?.Any() == true)
		? HttpStatusCode.BadRequest
		: HttpStatusCode.OK;

	string[] jsonArray = await Task.WhenAll(
		results.Select(r => _writer.WriteToStringAsync(r))
	);

	HttpResponseMessage response = Request.CreateResponse(httpResult);

	// I do not know what my HTTP response should be
	// `jsonArray` should be a string, but it is an array of JSON strings
	response.Content = new System.Net.Http.StringContent(jsonArray, Encoding.UTF8, "application/json");

	return response;
}
```

**Comment by furier** [comment]
any news on this?

**Comment by ZenSoftware** [comment]
I managed to get batching to work with [Apollo Http Batching Link](https://www.npmjs.com/package/apollo-angular-link-http-batch)!  Refer to this post for more information: [Query batching in Apollo](https://blog.apollographql.com/query-batching-in-apollo-63acfd859862).  I thought I'd post my graphql-dotnet code here that handles batched requests.  I am using the ASP.NET Core Middleware approach.  The solution is actually quite similar to how you would normally process a GQL request. You simply need to: 

1. Parse the incoming request as an `GraphQLRequest[]` instead of a `GraphQLRequest`
2. Process the batched requests into a resulting `ExecutionResult[]`.  
3. Return a JSON array containing the execution results, refer to `WriteResultsAsync()` function.

```C#
using GraphQL.Http;
using GraphQL.Types;
using Microsoft.AspNetCore.Antiforgery;
using Microsoft.AspNetCore.Authentication;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc.Formatters;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;

namespace GraphQL.Core
{
    public class GraphQLMiddleware
    {
        private readonly RequestDelegate _next;
        private readonly GraphQLSettings _settings;
        private readonly IDocumentExecuter _executer;
        private readonly IDocumentWriter _writer;

        private readonly static JsonSerializer _serializer = new JsonSerializer();
        private readonly static MediaType _jsonMediaType = new MediaType("application/json");

        public GraphQLMiddleware(
            RequestDelegate next,
            GraphQLSettings settings,
            IDocumentExecuter executer,
            IDocumentWriter writer)
        {
            _next = next;
            _settings = settings;
            _executer = executer;
            _writer = writer;
        }

        public async Task Invoke(HttpContext context, IAuthenticationService authenticationService, IAntiforgery antiforgery, IGraphQLEndpointProvider endpoint)
        {
            if (!endpoint.IsGraphQLRequest())
            {
                await _next(context);
            }
            else
            {
#if !DEBUG
                await antiforgery.ValidateRequestAsync(context);
#endif
                var principal = context.User;

                var authenticateResult = await authenticationService.AuthenticateAsync(context, "Api");

                if (authenticateResult.Succeeded)
                {
                    principal = authenticateResult.Principal;
                }
#if DEBUG
                var schema = endpoint.GetSchema();
                await ExecuteAsync(context, schema);
#else
                if(await endpoint.AuthorizeAsync(principal))
                {
                    var schema = endpoint.GetSchema();
                    await ExecuteAsync(context, schema);
                }
#endif
            }
        }

        private async Task ExecuteAsync(HttpContext context, ISchema schema)
        {
            GraphQLRequest[] requests = null;

            if (HttpMethods.IsPost(context.Request.Method))
            {

                var mediaType = new MediaType(context.Request.ContentType);

                try
                {
                    if (mediaType.IsSubsetOf(_jsonMediaType))
                    {
                        using (var sr = new StreamReader(context.Request.Body))
                        {
                            using (var jsonTextReader = new JsonTextReader(sr))
                            {
                                requests = _serializer.Deserialize<GraphQLRequest[]>(jsonTextReader);
                            }
                        }
                    }
                }
                catch (Exception e)
                {
                    await WriteErrorAsync(context, "An error occured while processing the GraphQL query", e);
                    return;
                }
            }

            ExecutionResult[] results = await Task.WhenAll(
                requests.Select(request => _executer.ExecuteAsync(options =>
                {
                    options.Schema = schema;
                    options.Query = request.Query;
                    options.OperationName = request.OperationName;
                    options.Inputs = request.Variables.ToInputs();
                    options.UserContext = _settings.BuildUserContext?.Invoke(context);
                    options.ValidationRules = _settings.ValidationRules;
#if DEBUG
                    options.ExposeExceptions = true;
#endif
                }))
            );

            HttpStatusCode httpResult = results.Any(result => result.Errors?.Any() == true)
                ? HttpStatusCode.BadRequest
                : HttpStatusCode.OK;
            
            context.Response.StatusCode = (int)httpResult;
            context.Response.ContentType = "application/json";
            await WriteResultsAsync(context, results);
        }

        private async Task WriteResultsAsync(HttpContext context, ExecutionResult[] results)
        {
            using (var sw = new StreamWriter(context.Response.Body, UTF8Encoding.UTF8))
            {
                sw.AutoFlush = true;
                sw.Write("[");
                for (int i = 0; i <= results.Length - 2; i++)
                {
                    await _writer.WriteAsync(context.Response.Body, results[i]);
                    sw.Write(",");
                }
                await _writer.WriteAsync(context.Response.Body, results.Last());
                sw.Write("]");
            }
        }

        private async Task WriteErrorAsync(HttpContext context, string message, Exception e = null)
        {
            if (message == null)
            {
                throw new ArgumentNullException(nameof(message));
            }

            var errorResult = new ExecutionResult();
            errorResult.Errors = new ExecutionErrors();

            if (e == null)
            {
                errorResult.Errors.Add(new ExecutionError(message));
            }
            else
            {
                errorResult.Errors.Add(new ExecutionError(message, e));
            }

            context.Response.StatusCode = (int)HttpStatusCode.BadRequest;
            context.Response.ContentType = "application/json";

            await _writer.WriteAsync(context.Response.Body, errorResult);
        }
    }

    public class GraphQLRequest
    {
        public string OperationName { get; set; }
        public string NamedQuery { get; set; }
        public string Query { get; set; }
        public JObject Variables { get; set; }
    }
}
```

Here is my `IGraphQLEndpointProvider` interface that I use to inject into the middleware above.
```C#
using GraphQL.Types;
using System.Security.Claims;
using System.Threading.Tasks;

namespace GraphQL.Core
{
    public interface IGraphQLEndpointProvider
    {
        ISchema GetSchema();
        Task<bool> AuthorizeAsync(ClaimsPrincipal principal);
        bool IsGraphQLRequest();
    }
}
```

**Comment by sungam3r** [comment]
Closed as requested feature was implemented in #241 . Feel free to reopen.
